### ISSUE TITLE
DisconnectedMessageBuffer design polluting disk and making persisted message non-recoverable as long as client process is running

### ISSUE THREAD

**Issue by rajeshnailwal** [eclipse-paho/paho.mqtt.java#759]
Please fill out the form below before submitting, thank you!

- [x] Bug exists Release Version 1.2.2 ( Master Branch)

Implementation of `public void sendNoWait(MqttWireMessage message, MqttToken token)` seems to have a basic flaw where invocation of `disconnectedMessageBuffer.putMessage` removes last message from in memory buffer in case deleteOldestMessages flag is set `bufferOpts.isDeleteOldestMessages() == true` 

`public void putMessage(MqttWireMessage message, MqttToken token) throws MqttException {`
		`BufferedMessage bufferedMessage = new BufferedMessage(message, token);`
		`synchronized (bufLock) {`
			`if (buffer.size() < bufferOpts.getBufferSize()) {`
				`buffer.add(bufferedMessage);`
			`} else if (bufferOpts.isDeleteOldestMessages() == true) {`
				`buffer.remove(0);`
				`buffer.add(bufferedMessage);`
			`} else {`
				`throw new MqttException(MqttException.REASON_CODE_DISCONNECTED_BUFFER_FULL);`
			`}`
		`}`
	`}` 

but the same message remains in persistence storage (in disk) and is never send for delivery as long as client is running. It seems to load such messages again one need to restart client (but this functionality is broken due to NullPointerException https://github.com/eclipse/paho.mqtt.java/issues/634).

This message should either be removed from disk or made available for delivery. I think removal is better option as in-memory buffer is designed for same.

**Comment by rajeshnailwal** [comment]
Is anyone looking on this issue?

**Comment by rdasgupt** [comment]
@rajeshnailwal 
Could you please try your test with the new service release 1.2.4?

**Comment by rajeshnailwal** [comment]
@rdasgupt I am trying to build but test case is stuck at some point every time for waiting for connection. However changes I saw are making some sense to me but I am still to verify it once build is created successfully

**Comment by rdasgupt** [comment]
Tests may take about 15 to 20 minutes to complete. Some of the tests (that try to connect to invalid host) may appear in hung state, but these tests should eventually timeout.

**Comment by rajeshnailwal** [comment]
This test case is stuck for 2 hours and do not complete -

20200425 180533.870 *************************************************************
20200425 180533.870 * ConnectionLossTest.testConnectionLossWhilePublishingQos1Async
20200425 180533.870 *************************************************************
20200425 180533.870 [CMPS Proxy] - Enabling Proxy
20200425 180533.870 [CMPS Proxy] - Waiting for incoming connection..
20200425 180533.870 Sat Apr 25 18:05:33 IST 2020 - Connecting...
20200425 180533.870 [CMPS Proxy] - Client Opened Connection to Proxy...
20200425 180535.893 ConnectionManipulationProxyServer cannot connect to localhost:1883
20200425 180535.893 [CMPS Proxy] - Waiting for incoming connection..
20200425 180535.893 [CMPS Proxy] - Client Opened Connection to Proxy...
20200425 180537.917 ConnectionManipulationProxyServer cannot connect to localhost:1883
20200425 180537.917 [CMPS Proxy] - Waiting for incoming connection..

**Comment by rdasgupt** [comment]
Are you trying to run tests with your changes in the repo?

**Comment by rajeshnailwal** [comment]
No, this is a fresh checkout of repo. I did not get chance to look into the test case. I am using gitbash to produce a maven build on windows.

**Comment by rdasgupt** [comment]
I will check on windows. Meanwhile if you need 1.2.4 library to try your application, you can use -DskipTests flag to bypass tests.
